pkg/lb/servicelb/manager.go

```diff
@@ -0,0 +1,295 @@
+	klog.Infof("uid: %s, isHealthy: %t", uid, isHealthy)
+	ns, name, server, err := unMarshalUID(uid)
+	if err != nil {
+		return err
+	}
+
+	eps, err := m.endpointSliceClient.Get(ns, name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	for i, ep := range eps.Endpoints {
+		if len(ep.Addresses) != 1 {
+			return fmt.Errorf("the length of addresses is not 1, endpoint: %+v", ep)
+		}
+		if ep.Addresses[0] == server {
+			epsCopy := eps.DeepCopy()
+			epsCopy.Endpoints[i].Conditions = discoveryv1.EndpointConditions{Ready: &isHealthy}
+			_, err := m.endpointSliceClient.Update(epsCopy)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *Manager) EnsureLoadBalancer(lb *lbv1.LoadBalancer) error {
+	if err := m.ensureService(lb); err != nil {
+		return err
+	}
+	if err := m.ensureEndpointSlice(lb); err != nil {
+		return err
+	}
+
+	return m.ensureProber(lb)
+}
+
+func (m *Manager) DeleteLoadBalancer(lb *lbv1.LoadBalancer) {
+	m.removeProber(lb)
+}
+
+func (m *Manager) ensureProber(lb *lbv1.LoadBalancer) error {
+	if lb.Spec.HeathCheck == nil {
+		m.removeProber(lb)
+		return nil
+	}
+	option := prober.HealthOption{
+		SuccessThreshold: lb.Spec.HeathCheck.SuccessThreshold,
+		FailureThreshold: lb.Spec.HeathCheck.FailureThreshold,
+		Timeout:          time.Duration(lb.Spec.HeathCheck.TimeoutSeconds) * time.Second,
+		Period:           time.Duration(lb.Spec.HeathCheck.PeriodSeconds) * time.Second,
+		InitialCondition: true,
+	}
+
+	eps, err := m.endpointSliceClient.Get(lb.Namespace, lb.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	existingEpUIDs := map[string]bool{}
+
+	for _, ep := range eps.Endpoints {
+		if len(ep.Addresses) != 1 {
+			return fmt.Errorf("the length of addresses is not 1, endpoint: %+v", ep)
+		}
+		option.Address = ep.Addresses[0] + ":" + strconv.Itoa(lb.Spec.HeathCheck.Port)
+		uid := marshalUID(lb.Namespace, lb.Name, ep.Addresses[0])
+		existingEpUIDs[uid] = true
+		if ep.Conditions.Ready != nil {
+			option.InitialCondition = *ep.Conditions.Ready
+		}
+		m.AddWorker(uid, option)
+	}
+
+	// remove the outdated workers
+	workerMap := m.ListWorkers()
+	for uid := range workerMap {
+		if !existingEpUIDs[uid] {
+			m.RemoveWorker(uid)
+		}
+	}
+
+	return nil
+}
+
+func (m *Manager) removeProber(lb *lbv1.LoadBalancer) {
+	for _, server := range lb.Spec.BackendServers {
+		uid := marshalUID(lb.Namespace, lb.Name, server)
+		m.RemoveWorker(uid)
+	}
+}
+
+func unMarshalUID(uid string) (namespace, name, server string, err error) {
+	fields := strings.Split(uid, "/")
+	if len(fields) != 3 {
+		err = fmt.Errorf("invalid uid %s", uid)
+		return
+	}
+	namespace, name, server = fields[0], fields[1], fields[2]
+	return
+}
+
+func marshalUID(ns, name, server string) (uid string) {
+	uid = ns + "/" + name + "/" + server
+	return
+}
+
+func (m *Manager) ensureService(lb *lbv1.LoadBalancer) error {
+	svc, err := m.serviceClient.Get(lb.Namespace, lb.Name, metav1.GetOptions{})
```

Comment:
Can we get from cache.

Reply:
No, because we have no informer of this object.

Reply:
I mean we should use cache